### PR TITLE
Add __array_func__ and np.linspace support for astropy.time.Time

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
This is a complete re-release of @dstansby's #12703, rebased and adding more test coverage and documentation.

xref #8610

close #12703

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/time/time_helper/function_helpers.py
Comment (pllim) [inline astropy/time/time_helper/function_helpers.py:9]: Does this TODO need to be addressed before this can be merged? If not, do we need a follow-up issue?

File: docs/time/index.rst
Comment (pllim) [inline docs/time/index.rst:155]:
```suggestion
  >>> np.linspace(t, t + [4, 120] * u.day, 4, endpoint=False)
```

File: docs/time/index.rst
Comment (pllim) [inline docs/time/index.rst:151]:
```suggestion
As of v5.1, |Time| objects can also be passed directly to
```

**Comment by pllim** [comment]
Seems like a pretty cool enhancement but I'll let subpackage maintainer(s) decide if What's New entry is needed or not.

**Comment by pllim** [comment]
Why not also run the "exotic" archs jobs while we wait for review, just in case...

File: docs/time/index.rst
Comment (mhvk) [reply docs/time/index.rst:155]: Small point, but ideally we give an example with actual `Time` instances for both start and stop.

**Review by mhvk — COMMENTED** [review]
I think this looks good, but am wondering if it isn't better to merge this for 5.2 (i.e., after upcoming 5.1 deadline), so that we can have an explicit issue about dealing with unsupported functions as well as possibly supporting others (`np.isclose`, etc.).

But am also OK with getting this in so that we get started.

File: astropy/time/time_helper/function_helpers.py
Comment (dhomeier) [reply astropy/time/time_helper/function_helpers.py:9]: Good question – not quite sure what was intended to add there in #12703, but assuming it was to add some sort of graceful exit when calling a numpy function that cannot return sensible results for `Time`.
So I'd expect as long as this is unfilled, those functions will just continue to behave the same they are already doing now, so assuming it does not need to be completed right now.

**Comment by dhomeier** [comment]
Since it is an enhancement, I'd like to get it into 5.1 if possible, but of course not if it seems incomplete.
The code was only changed in one point from the implementation that was up for final review of #12703, to allow handling calls to `np.linspace(..., retstep=True)`.
I think `linspace` is a quite nice convenience to start with; initially I became aware of this in context of the huge performance bug fixed in #13069, as this will allow `np.searchsorted` calls on `Time` objects to be fixed in more generality, and that would be the next custom function to be added in a follow-up.

**Comment by dhomeier** [comment]
> Seems like a pretty cool enhancement but I'll let subpackage maintainer(s) decide if What's New entry is needed or not.

I've recycled the one from the original PR just in case.

**Comment by mhvk** [comment]
OK, let's go for 5.1 then, I'm probably being overly cautious and this is really neat. But could you add a test somewhere of `retstep`? I think we'd want to check that it is a `TimeDelta` and has the right value/shape (does not have to be for all `Time` formats).

**Comment by dhomeier** [comment]
The values are checked in `test_linspace_steps`, but not explicitly the type; just testing for the expected `TimeDelta`.

**Comment by mhvk** [comment]
Duh, don't know how I missed it. Maybe failed to refresh...

**Review by mhvk — APPROVED** [review]
OK, I think you've got it all covered. Let's get this in! (once tests pass, of course)

**Comment by pllim** [comment]
Thanks, all!

**Comment by dhomeier** [comment]
Thanks for reviewing and input, and thanks to @dstansby for implementing this in the first place!

**Pull request merged**
